### PR TITLE
fix: precision infinite loop bug solution by edge casing calculations

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

This a solution to the bug stemming from the difference between sqrt prices being truncated to zero due to lack of precision.

The difference of zero leads to computing amounts in/out in a swap step with the value of zero. In turn, this results in the inability to progress in the swap loop. The progress can only be made if we keep subtracting computed in/out values from the `amountIn/OutRemaining`

An alternative solution is to increase precision for intermediary calculations like in this PR: https://github.com/osmosis-labs/osmosis/pull/5612

Here, we focus on detecting when incorrect truncation occurs and setting the amountIn/Out to the amount remaining so that the swap makes progress.

Detailed comments are left directly in the code to explain the logic.

## Testing and Verifying

Uses Alpin's repro + added unit tests reproducing this.

Each added compute swap step logic branch is hit at least once by a direct unit test.

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [x] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/concentrated-liquidity/swapstrategy/one_for_zero.go [hunk 1]
```diff
@@ -87,6 +87,23 @@ func (s oneForZeroStrategy) ComputeSwapWithinBucketOutGivenIn(sqrtPriceCurrent,
...
+	// This covers an edge case where due to the lack of precision, the difference between the current sqrt price and the next sqrt price is so small that
+	// it ends up being rounded down to zero. This leads to an infinite loop in the swap algorithm. From knowing that this is a case where !hasReachedTarget,
+	//(that is the swap stops within a bucket), we charge the full amount remaining in to the user and infer the amount out from the sqrt price truncated
+	// in favor of the pool.
+	if !hasReachedTarget && sqrtPriceCurrent.Equal(sqrtPriceNext) && amountOneIn.IsZero() && !amountOneInRemaining.IsZero() {
+		amountOneIn = amountOneInRemaining
+		// Subtract 1 ULP so that the amount out is rounded in favor of the pool.
+		// TODO: figure out precision difference in the denominations of two tokens,
+		// leading to incorrect spot price in out core logic.
+		amountZeroOut = sqrtPriceCurrent.Sub(oneULP).PowerMut(2)
```
Comment (p0mvn) [inline x/concentrated-liquidity/swapstrategy/one_for_zero.go:104]: I'm referring to the issue described [here](https://github.com/osmosis-labs/osmosis/blob/83a14999fa5cc813f9dbc2f763b3917044ef4c5a/x/concentrated-liquidity/README.md?plain=1#L1628-L1638)

File: x/concentrated-liquidity/swapstrategy/one_for_zero.go [hunk 2]
```diff
@@ -87,6 +87,23 @@ func (s oneForZeroStrategy) ComputeSwapWithinBucketOutGivenIn(sqrtPriceCurrent,
...
+	// This covers an edge case where due to the lack of precision, the difference between the current sqrt price and the next sqrt price is so small that
+	// it ends up being rounded down to zero. This leads to an infinite loop in the swap algorithm. From knowing that this is a case where !hasReachedTarget,
+	//(that is the swap stops within a bucket), we charge the full amount remaining in to the user and infer the amount out from the sqrt price truncated
+	// in favor of the pool.
+	if !hasReachedTarget && sqrtPriceCurrent.Equal(sqrtPriceNext) && amountOneIn.IsZero() && !amountOneInRemaining.IsZero() {
+		amountOneIn = amountOneInRemaining
+		// Subtract 1 ULP so that the amount out is rounded in favor of the pool.
+		// TODO: figure out precision difference in the denominations of two tokens,
+		// leading to incorrect spot price in out core logic.
+		amountZeroOut = sqrtPriceCurrent.Sub(oneULP).PowerMut(2).MulTruncate(amountOneIn)
```
Comment (p0mvn) [inline x/concentrated-liquidity/swapstrategy/one_for_zero.go]: On more thinking, I don't think the precision difference would be an issue here. Although the spot price, would be incorrect due to the precision difference it does not affect the correctness of this calculation.

Assume:
- quote asset (token 1) of uosmo with precision 6 and reserves `218023341414`
- base asset (token 0) of DAI with precision 18 and reserves `101170077995723619690981`
- `amountOneIn` of `1_000_000_000_000`

Therefore, the true price of the pool is:
```python
>>> (218023341414 / 10**6)  / (101170077995723619690981 / 10**18)
2.1550180224553714
```

However, in our core logic, it is represented as:

```python
218023341414 / 101170077995723619690981
2.1550180224553714e-12
```

Therefore, in our logic here, we would calculate:
```
1 / 2.1550180224553714e-12 = 464033242218.8405

100 * 464033242218.8405 = 46403324221884.055
```

When the true value is:
```
1 / 2.1550180224553714 = 0.46403324221884046

100 * 0.46403324221884046 =46.403324221884046
```

File: x/concentrated-liquidity/swapstrategy/one_for_zero.go [hunk 3]
```diff
@@ -87,6 +87,31 @@ func (s oneForZeroStrategy) ComputeSwapWithinBucketOutGivenIn(sqrtPriceCurrent,
...
+		// leading to incorrect spot price in out core logic.
+
+		// sqrt price is amount 1 / amount 0
+		// How many tokens 1 do we need to get 1 token 0?
+		priceOneOverZero := sqrtPriceCurrent.MulRoundUp(sqrtPriceCurrent)
+
+		// how many tokens 0 do we get for 1 token 1?
+		priceZeroOverOne := sdk.OneDec().QuoTruncate(priceOneOverZero)
+
+		amountZeroOut = amountOneIn.Mul(priceZeroOverOne)
```
Comment (p0mvn) [inline x/concentrated-liquidity/swapstrategy/one_for_zero.go:112]: On more thinking, I don't think the precision difference would be an issue here. Although the spot price, would be incorrect due to the precision difference it does not affect the correctness of this calculation.

Assume:
- quote asset (token 1) of uosmo with precision 6 and reserves `218023341414`
- base asset (token 0) of DAI with precision 18 and reserves `101170077995723619690981`
- `amountOneIn` of `1_000_000_000_000`

Therefore, the true price of the pool is:
```python
>>> (218023341414 / 10**6)  / (101170077995723619690981 / 10**18)
2.1550180224553714
```

However, in our core logic, it is represented as:

```python
218023341414 / 101170077995723619690981
2.1550180224553714e-12
```

Therefore, in our logic here, we would calculate:
```
1 / 2.1550180224553714e-12 = 464033242218.8405

100 * 464033242218.8405 = 46403324221884.055
```

When the true value is:
```
1 / 2.1550180224553714 = 0.46403324221884046

100 * 0.46403324221884046 =46.403324221884046
```

This means that we would remove `46403324221884` from reserves that is in fact: `46403324221884 / 10**12 = 46.403324221884046`

File: x/concentrated-liquidity/swapstrategy/one_for_zero.go [hunk 4]
```diff
@@ -83,10 +84,40 @@ func (s oneForZeroStrategy) ComputeSwapWithinBucketOutGivenIn(sqrtPriceCurrent,
...
+		// How many tokens 1 do we need to get 1 token 0?
+
+		priceOneOverZero := sqrtPriceCurrent.MulRoundUp(sqrtPriceCurrent)
+
+		// how many tokens 0 do we get for 1 token 1?
+		priceZeroOverOne := osmomath.OneDec().QuoTruncate(osmomath.BigDecFromSDKDec(priceOneOverZero))
+
+		if priceZeroOverOne.IsZero() {
+			panic("priceZeroOverOne is zero")
+		}
```
Comment (p0mvn) [inline x/concentrated-liquidity/swapstrategy/one_for_zero.go:107]: This panic can easily be hit

File: osmomath/decimal_test.go [hunk 5]
```diff
@@ -356,40 +356,40 @@ func (s *decimalTestSuite) TestDecsEqual() {
 func (s *decimalTestSuite) TestArithmetic() {
 	tests := []struct {
 		d1, d2                                osmomath.BigDec
-		expMul, expMulTruncate                osmomath.BigDec
+		expMul, expMulTruncate, expMulRoundUp osmomath.BigDec
```
Comment (p0mvn) [inline osmomath/decimal_test.go:359]: Note to reviewer: all tests in this file are copied from: https://github.com/osmosis-labs/cosmos-sdk/pull/430

File: x/concentrated-liquidity/math/math.go [hunk 6]
```diff
@@ -132,11 +170,12 @@ func GetNextSqrtPriceFromAmount0InRoundingUp(sqrtPriceCurrent, liquidity, amount
...
 	}
 
-	product := amountZeroRemainingIn.Mul(sqrtPriceCurrent)
+	// MulTruncate to get a smaller denominator
+	product := amountZeroRemainingIn.MulTruncate(sqrtPriceCurrent)
 	// denominator = product + liquidity
 	denominator := product
 	denominator.AddMut(liquidity)
-	return liquidity.Mul(sqrtPriceCurrent).QuoRoundupMut(denominator)
+	return liquidity.MulRoundUp(sqrtPriceCurrent).QuoRoundupMut(denominator)
```
Comment (p0mvn) [inline x/concentrated-liquidity/math/math.go:178]: Note: the reason why this change was made is to ensure that if any rounding error occurs, the value computed by this function is always greater than the true value per requirements. With the old change, banker's rounding can break this guarantee.

No test ended up breaking from this change.

File: x/concentrated-liquidity/math/math.go [hunk 7]
```diff
@@ -90,6 +90,44 @@ func CalcAmount0Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec, roundUp bool) sdk.Dec
...
 }
 
+// CalcAmount0DeltaBigDec takes the asset with the smaller liquidity in the pool as well as the sqrtpCur and the nextPrice and calculates the amount of asset 0
+// sqrtPriceA is the smaller of sqrtpCur and the nextPrice
+// sqrtPriceB is the larger of sqrtpCur and the nextPrice
+// CalcAmount0Delta = (liquidity * (sqrtPriceB - sqrtPriceA)) / (sqrtPriceB * sqrtPriceA)
+// Note, that the only reason we have a BigDec version of this function is to account for an edge case that arises due to lack of precision
+// in one for zero swap strategy's ComputeSwapWithinBucketOutGivenIn function. See the comment in that function
+// for more details.
+func CalcAmount0DeltaBigDec(liq, sqrtPriceA, sqrtPriceB osmomath.BigDec, roundUp bool) osmomath.BigDec {
```
Comment (p0mvn) [inline x/concentrated-liquidity/math/math.go:100]: Note: this is a 1:1 copy of the `sdk.Dec` version `CalcAmount0Delta`

File: x/concentrated-liquidity/math/math.go [hunk 8]
```diff
@@ -163,6 +202,18 @@ func GetNextSqrtPriceFromAmount1InRoundingDown(sqrtPriceCurrent, liquidity, amou
...
 
+// GetNextSqrtPriceFromAmount1InRoundingDownBigDec utilizes the current sqrtPriceCurrent, liquidity, and amount of denom1 that still needs
+// to be swapped in order to determine the sqrtPriceNext.
+// When we swap for token zero out given token one in, the price is increasing and we need to move the sqrt price (increase it) less to
+// avoid overpaying out of the pool. Therefore, we round down.
+// sqrt_next = sqrt_cur + token_in / liq
+// Note, that the only reason we have a BigDec version of this function is to account for an edge case that arises due to lack of precision
+// in one for zero swap strategy's ComputeSwapWithinBucketOutGivenIn function. See the comment in that function
+// for more details.
+func GetNextSqrtPriceFromAmount1InRoundingDownBigDec(sqrtPriceCurrent, liquidity, amountOneRemainingIn osmomath.BigDec) (sqrtPriceNext osmomath.BigDec) {
```
Comment (p0mvn) [inline x/concentrated-liquidity/math/math.go:213]: Note: 1:1 copy of the `sdk.Dec` version

File: x/concentrated-liquidity/math/math_test.go [hunk 9]
```diff
@@ -156,6 +162,37 @@ func (suite *ConcentratedMathTestSuite) TestGetNextSqrtPriceFromAmount1RoundingD
 	}
 }
 
+// TestGetNextSqrtPriceFromAmount1RoundingDownBigDec tests that getNextSqrtPriceFromAmount1RoundingDown
+// utilizes the current squareRootPrice, liquidity of denom1, and amount of denom1 that still needs
+// to be swapped in order to determine the next squareRootPrice
+// sqrtPriceNext = sqrtPriceCurrent + (amount1Remaining / liquidity1)
+func (suite *ConcentratedMathTestSuite) TestGetNextSqrtPriceFromAmount1RoundingDownBigDec() {
```
Comment (p0mvn) [inline x/concentrated-liquidity/math/math_test.go:169]: Note: both tests added in this file are 1:1 copies with final results recalculated in Python

File: osmomath/decimal.go [hunk 10]
```diff
@@ -305,6 +305,17 @@ func (d BigDec) MulTruncate(d2 BigDec) BigDec {
 	return BigDec{chopped}
 }
 
+// multiplication round up
+func (d BigDec) MulRoundUp(d2 BigDec) BigDec {
```
Comment (ValarDragon) [inline osmomath/decimal.go:309]: wait how does this ever affect anything?

Multiplication has no rounding under decimal?

File: x/concentrated-liquidity/swapstrategy/one_for_zero.go [hunk 11]
```diff
@@ -83,6 +84,24 @@ func (s oneForZeroStrategy) ComputeSwapWithinBucketOutGivenIn(sqrtPriceCurrent,
 	// Calculate the amount of the other token given the sqrt price range.
 	amountZeroOut := math.CalcAmount0Delta(liquidity, sqrtPriceNext, sqrtPriceCurrent, false)
 
+	// This covers an edge case where due to the lack of precision, the difference between the current sqrt price and the next sqrt price is so small that
+	// it ends up being rounded down to zero. This leads to an infinite loop in the swap algorithm. From knowing that this is a case where !hasReachedTarget,
+	// (that is, the swap stops within a bucket), we charge the full amount remaining in to the user and recalculate the amount out with higher precision.
+	// Notice, that in the zero for one direction we instead use the sqrtPriceCurrent to calculate the current price to get the amount out from amount in.
+	// We cannot do that here because that would require us inverting the current price. Due to our internal fixed point math, inverting it would lead to a price of zero,
+	// breaking the calculation. As a result, we recalculate the amount out in sqrt price domain with higher precision here.
+	if !hasReachedTarget && sqrtPriceCurrent.Equal(sqrtPriceNext) && amountOneIn.IsZero() && !amountOneInRemaining.IsZero() {
```
Comment (ValarDragon) [inline x/concentrated-liquidity/swapstrategy/one_for_zero.go:93]: Great job on this comment.

So for safety we have to convince ourselves that this case could never cause us to actually cross a tick in this higher precision domain.

This makes sense if we believe the claim that we can only enter this case when `sqrtPriceReached = sqrtPriceCurrent` in the lower precision representation, and `sqrtPriceCurrent < sqrtPriceTarget`. 

And importantly that under big dec, the sqrtPriceReached still can never hit sqrtPriceTarget (due to rounding somewhere else)

File: x/concentrated-liquidity/swapstrategy/one_for_zero_test.go [hunk 12]
```diff
@@ -148,6 +148,61 @@ func (suite *StrategyTestSuite) TestComputeSwapStepOutGivenIn_OneForZero() {
...
+
+			expectedAmountInConsumed: sdk.NewDec(99),
+			// liquidity * (sqrtPriceNext - sqrtPriceCurrent) / (sqrtPriceNext * sqrtPriceCurrent)
+			// calculated with x/concentrated-liquidity/python/clmath.py
+			// diff = (sqrtPriceNext - sqrtPriceCurrent)
+			// diff = round_decimal(diff, 36, ROUND_FLOOR) (0.000000000000000000989975269800000000)
+			// mul = (sqrtPriceNext * sqrtPriceCurrent)
+			// mul = round_decimal(mul, 36, ROUND_CEILING) (0.000000000001000100000000865026329827)
+			//  round_decimal(liquidity * diff / mul, 36, ROUND_FLOOR)
+			expectedAmountOut:               osmomath.MustNewDecFromStr("98990100989815.389417309844929293132374729779331247").SDKDec(),
```
Comment (p0mvn) [inline x/concentrated-liquidity/swapstrategy/one_for_zero_test.go:187]: large out

Comment (p0mvn) [reply x/concentrated-liquidity/swapstrategy/one_for_zero_test.go:187]:
```
 (1 / Decimal("0.000001000049998751")) * 99
Decimal('98995050.3711262616004566779050751370')
```

Comment (p0mvn) [reply x/concentrated-liquidity/swapstrategy/one_for_zero_test.go:187]:
```
(1 / (Decimal("0.000001000049998751") * Decimal("0.000001000049998751"))) * 99
Decimal('98990100989715.4119660129268042850683')
```

Comment (ValarDragon) [reply x/concentrated-liquidity/swapstrategy/one_for_zero_test.go:187]: This checks out, 99 units are a sub-ulp swap, over 10^-12 spot price ~= 99 * 10^12

**Comment by ValarDragon** [comment]
Per discussion on call, we were wrong when we initially planned these approaches. We have to swiitch to a different approach.

**Comment by p0mvn** [comment]
Approach: https://hackmd.io/6wKtoyK6TPS6Bxe_AiHmCw
